lib/services/protocol_service.ts

```diff
@@ -736,41 +736,42 @@ export class SNProtocolService extends PureService implements EncryptionDelegate
    * @returns JSON stringified representation of data, including keyParams.
    */
   public async createBackupFile(
-    subItems?: SNItem[],
-    intent = EncryptionIntent.FilePreferEncrypted,
-    returnIfEmpty = false
-  ) {
-    const items = subItems || this.itemManager!.items;
-    if (returnIfEmpty && items.length === 0) {
-      return undefined;
-    }
-    const encryptedPayloads: PurePayload[] = [];
-    for (const item of items) {
-      if (item.errorDecrypting) {
-        /** Keep payload as-is */
-        encryptedPayloads.push(item.payload);
-      } else {
-        const payload = CreateSourcedPayloadFromObject(
-          item.payload,
-          PayloadSource.FileImport
-        );
-        const encrypted = await this.payloadByEncryptingPayload(
-          payload,
-          intent
-        );
-        encryptedPayloads.push(encrypted);
-      }
+    intent: EncryptionIntent
+  ): Promise<BackupFile> {
+    let items = this.itemManager.items;
+
+    if (intent === EncryptionIntent.FileDecrypted) {
+      items = items.filter(item => item.content_type !== ContentType.ItemsKey);
     }
+
+    const ejectedPayloads = Promise.all(
+      items.map(item => {
+        if (item.errorDecrypting) {
+          /** Keep payload as-is */
+          return item.payload.ejected();
+        } else {
+          const payload = CreateSourcedPayloadFromObject(
+            item.payload,
+            PayloadSource.FileImport
+          );
+          return this.payloadByEncryptingPayload(
+            payload,
+            intent
+          ).then(p => p.ejected());
+        }
+      })
+    );
+
+    const keyParams = await this.getRootKeyParams();
+
     const data: BackupFile = {
       version: this.getLatestVersion(),
-      items: encryptedPayloads.map((p) => p.ejected())
+      items: await ejectedPayloads,
```

Comment:
Why not `await` at the definition level instead? In this case the `ejectedPayloads` variable is misnamed, as it's a promise, not ejectedVariables. Seems like the await should be at line 747.

I don't think there's any performance optimization to be gained here either from postponing the await. JS is ultimately single-threaded.

Reply:
@radko93 has reported notable performance improvements from "parallel" processing. 

While JS is single-threaded, when an async function is awaiting on a disk, network, or encryption result (like when you call into the RN bridge), it's put in a queue and other JS sync code can be run in the meantime. This is why using things like Promise.all, while not fully parallelizing the work, schedules it smarter, resulting in better performance. Does that make sense?

All that being said, in this instance `getRootKeyParams` doesn't seem to be a heavy operation, it just comes at the tail-end of a very heavy one (depending on how many notes/tags you have), so I went with as much smart scheduling as possible. If you think that makes it hard to read let me know and I'll `await` at the definition instead like you proposed.